File: frontend/src/types/writing-session.ts
/** AI 导向写作会话的前端类型定义（与后端 writing_session schema 对齐）。 */

export type WritingStage =
  | 'clarifying'
  | 'outline_review'
  | 'drafting'
  | 'draft_review'
  | 'editing'
  | 'completed';

export type WritingStatus = 'active' | 'completed' | 'abandoned';

/** 会话内的对话消息。 */
export interface WritingMessage {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  created_at: string;
}

/** AI 改稿建议。 */
export interface WritingSuggestion {
  id: string;
  type: 'structure' | 'argument' | 'readability' | 'seo' | 'accuracy';
  title: string;
  reason: string;
  scope: string;
  status: 'pending' | 'previewed' | 'applied' | 'dismissed';
}

/** 单次改稿修订（来源：选区指令 / 建议应用）。 */
export interface WritingRevision {
  id: string;
  source: 'selection' | 'suggestion';
  suggestion_id: string | null;
  content_hash: string;
  selection_start: number;
  selection_end: number;
  original_text: string;
  replacement_text: string;
  status: 'previewed' | 'applied' | 'discarded';
}

/** 写作会话聚合根。 */
export interface WritingSession {
  id: string;
  user_id: string;
  article_id: string | null;
  stage: WritingStage;
  status: WritingStatus;
  requirements_summary: Record<string, string>;
  outline: string;
  draft: string;
  messages: WritingMessage[];
  suggestions: WritingSuggestion[];
  revisions: WritingRevision[];
  created_at: string;
  updated_at: string;
}

/** 基于正文选区的改稿请求体（revise-selection/stream）。 */
export interface WritingSelectionRevisionRequest {
  content: string;
  selected_text: string;
  selection_start: number;
  selection_end: number;
  instruction: string;
  content_hash: string;
}

/** 基于建议的改稿请求体（revise-suggestion/stream）。 */
export interface WritingSuggestionRevisionRequest {
  suggestion_id: string;
  content: string;
  content_hash: string;
}

File: backend/app/schemas/writing_session.py
from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_serializer

WritingStage = Literal["clarifying", "outline_review", "drafting", "draft_review", "editing", "completed"]
WritingStatus = Literal["active", "completed", "abandoned"]


class WritingMessage(BaseModel):
    id: str
    role: Literal["user", "assistant", "system"]
    content: str
    created_at: datetime


class WritingSuggestion(BaseModel):
    id: str
    type: Literal["structure", "argument", "readability", "seo", "accuracy"]
    title: str
    reason: str
    scope: str
    status: Literal["pending", "previewed", "applied", "dismissed"] = "pending"


class WritingSessionCreate(BaseModel):
    article_id: Optional[UUID] = None


class WritingSessionRead(BaseModel):
    id: UUID
    user_id: UUID
    article_id: Optional[UUID]
    stage: WritingStage
    status: WritingStatus
    requirements_summary: dict[str, str]
    outline: str
    draft: str
    messages: list[WritingMessage]
    suggestions: list[WritingSuggestion]
    revisions: list[dict]
    created_at: datetime
    updated_at: datetime

    @field_serializer("id", "user_id", "article_id")
    def serialize_uuid(self, value: Optional[UUID]):
        return str(value) if value else None

    model_config = {"from_attributes": True}


class WritingMessageRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=4000)


class WritingOutlineAdjustRequest(WritingMessageRequest):
    pass


class WritingDraftAdjustRequest(WritingMessageRequest):
    pass


class WritingAnalyzeRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=30000)
    content_hash: str = Field(..., min_length=16, max_length=128)


class WritingSelectionRevisionRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=30000)
    selected_text: str = Field(..., min_length=1, max_length=10000)
    selection_start: int = Field(..., ge=0)
    selection_end: int = Field(..., ge=1)
    instruction: str = Field(..., min_length=1, max_length=1000)
    content_hash: str = Field(..., min_length=16, max_length=128)


class WritingSuggestionRevisionRequest(BaseModel):
    suggestion_id: str
    content: str = Field(..., min_length=1, max_length=30000)
    content_hash: str = Field(..., min_length=16, max_length=128)


class WritingRevisionApplyRequest(BaseModel):
    revision_id: str
    content_hash: str = Field(..., min_length=16, max_length=128)


class WritingRevisionDiscardRequest(BaseModel):
    revision_id: str


class WritingArticleLinkRequest(BaseModel):
    article_id: UUID
